src/sentry/processing/backpressure/memory.py

```diff
@@ -0,0 +1,64 @@
+from dataclasses import dataclass
+from typing import Any, Generator, Mapping, Union
+
+import requests
+from redis import Redis
+from rediscluster import RedisCluster
+
+
+@dataclass
+class ServiceMemory:
+    used: int
+    available: int
+    percentage: float
+
+    def __init__(self, used: int, available: int):
+        self.used = used
+        self.available = available
+        self.percentage = used / available
+
+
+def query_rabbitmq_memory_usage(host: str) -> ServiceMemory:
+    """Returns the currently used memory and the memory limit of a
+    RabbitMQ host.
+    """
+
+    if not host.endswith("/"):
+        host += "/"
+    url = f"{host}api/nodes"
+
+    response = requests.get(url)
+    response.raise_for_status()
+    json = response.json()
+    return ServiceMemory(json[0]["mem_used"], json[0]["mem_limit"])
+
+
+# Based on configuration, this could be:
+# - a `rediscluster` Cluster (actually `RetryingRedisCluster`)
+# - a straight `Redis` client (actually `FailoverRedis`)
+# - or any class configured via `client_class`.
+# It could in theory also be a `rb` (aka redis blaster) Cluster, but we
+# intentionally do not support these.
+Cluster = Union[RedisCluster, Redis]
+
+
+def get_memory_usage(info: Mapping[str, Any]) -> ServiceMemory:
+    # or alternatively: `used_memory_rss`?
+    memory_used = info.get("used_memory", 0)
+    # `maxmemory` might be 0 in development
+    memory_available = info.get("maxmemory", 0) or info.get("total_system_memory", 0)
+
+    return ServiceMemory(memory_used, memory_available)
```

Comment:
This can now throw a `ZeroDivisionError`, since internally you unconditionally divide by a number that defaults to `0`. Would it make sense to add more checks at some point? See the [original comment](https://github.com/getsentry/sentry/pull/50928/#discussion_r1230956425) for suggestions.

If `"total_system_memory"` is always there and non-zero, then you could just use direct field access to show clear intent.